tls/s2n_client_extensions.c

```diff
@@ -263,60 +268,38 @@ static int s2n_recv_client_server_name(struct s2n_connection *conn, struct s2n_s
     return 0;
 }
 
-int s2n_send_client_signature_algorithms(struct s2n_stuffer *out)
+static int s2n_recv_client_signature_algorithms(struct s2n_connection *conn, struct s2n_stuffer *extension)
 {
-    /* The array of hashes and signature algorithms we support */
-    uint16_t preferred_hashes_len = sizeof(s2n_preferred_hashes) / sizeof(s2n_preferred_hashes[0]);
-    uint16_t preferred_hashes_size = preferred_hashes_len * 2;
-    GUARD(s2n_stuffer_write_uint16(out, preferred_hashes_size));
-
-    for (int i =  0; i < preferred_hashes_len; i++) {
-        GUARD(s2n_stuffer_write_uint8(out, s2n_preferred_hashes[i]));
-        GUARD(s2n_stuffer_write_uint8(out, TLS_SIGNATURE_ALGORITHM_RSA));
-    }
-    return 0;
+    return s2n_recv_supported_signature_algorithms(conn, extension, &conn->handshake_params.client_sig_hash_algs);
 }
-
-int s2n_recv_client_signature_algorithms(struct s2n_connection *conn, struct s2n_stuffer *in, s2n_hash_algorithm *hash, s2n_signature_algorithm *sig)
+    
+int s2n_get_supported_signature_hash_pair(struct s2n_stuffer *in, s2n_hash_algorithm *hash_alg, s2n_signature_algorithm *signature_alg)
 {
-    uint16_t length_of_all_pairs;
-    GUARD(s2n_stuffer_read_uint16(in, &length_of_all_pairs));
-    if (length_of_all_pairs > s2n_stuffer_data_available(in)) {
-        /* Malformed length, ignore the extension */
-        return 0;
-    }
+    uint8_t hash_algorithm;
+    uint8_t signature_algorithm;
 
-    if (length_of_all_pairs % 2 || s2n_stuffer_data_available(in) % 2) {
-        /* Pairs occur in two byte lengths. Malformed length, ignore the extension. */
-        return 0;
+    GUARD(s2n_stuffer_read_uint8(in, &hash_algorithm));
+    GUARD(s2n_stuffer_read_uint8(in, &signature_algorithm));
+    
+    if (signature_algorithm != TLS_SIGNATURE_ALGORITHM_RSA && signature_algorithm != TLS_SIGNATURE_ALGORITHM_ECDSA) {
+        S2N_ERROR(S2N_ERR_INVALID_SIGNATURE_ALGORITHM);
     }
```

Comment:
Is this going to be too brittle? Especially in client mode ... what happens when folks go add some new signature type in the future?